nucypher/characters/lawful.py

```diff
@@ -563,36 +581,48 @@ def construct_hrac_and_map_id(self, verifying_key, label):
         map_id = keccak_digest(bytes(verifying_key) + hrac).hex()
         return hrac, map_id
 
-    def get_treasure_map_from_known_ursulas(self, network_middleware, map_id):
+    def get_treasure_map_from_known_ursulas(self, network_middleware, map_id, timeout=3):
         """
         Iterate through the nodes we know, asking for the TreasureMap.
         Return the first one who has it.
         """
-        from nucypher.policy.collections import TreasureMap
-        for node in self.known_nodes.shuffled():
-            try:
-                response = network_middleware.get_treasure_map_from_node(node=node, map_id=map_id)
-            except NodeSeemsToBeDown:
-                continue
-            except network_middleware.NotFound:
-                self.log.info(f"Node {node} claimed not to have TreasureMap {map_id}")
-                continue
-
-            if response.status_code == 200 and response.content:
+        if self.federated_only:
+            from nucypher.policy.collections import TreasureMap as _MapClass
+        else:
+            from nucypher.policy.collections import DecentralizedTreasureMap as _MapClass
+
+        start = maya.now()
+
+        # Spend no more than half the timeout finding the nodes.  8 nodes is arbitrary.  Come at me.
+        self.block_until_number_of_known_nodes_is(8, timeout=timeout/2, learn_on_this_thread=True)
+        while True:
+
+            nodes_with_map = self.matching_nodes_among(self.known_nodes)
+            random.shuffle(nodes_with_map)
```

Comment:
If we change this just `shuffle` we can remove `import random` at the top of the file.

Reply:
I prefer to keep this one and change the other to `random.shuffle` in the invocation.